### PR TITLE
WIP: milk-planning subdomain

### LINKED ISSUE #49: Milk planning - Define ubiquitous language

### LINKED ISSUE #51: Milk planning - Define incoming and outgoing events

### PR DESCRIPTION
**Pull request by nicolasfara**
Closes #49 
Closes #51

### CODE REVIEW

File: docs/_docs/milk-planning.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+# Milk Planning
+
+Every Saturday Raffaella has to estimate the quintals of milk necessary to produce all products
+for the following week.
+She makes this estimate by taking into account the following factors:
+
+- the quintals of milk processed in the same period of the previous year
+- the quintals of milk needed by the products that have to be produced in the following week
+  (this is made reading from a recipe book the the quintals of milk needed to produce a quintal of a given product)
+- the current stock
```
Comment (vitlinda) [inline docs/_docs/milk-planning.md:14]: Remove this
```suggestion
```

**Review by vitlinda — CHANGES_REQUESTED** [review]

**Review by vitlinda — CHANGES_REQUESTED** [review]
Remove the `Actions.scala`, `Events.scala` and `Types.scala` from the `types` package and place it simply in the `milkplanning` one as it's done in the other projects.

File: docs/_docs/milk-planning.md (hunk 1 above)
Comment (nicolasfara) [reply docs/_docs/milk-planning.md:14]: We re-introduced the current stock as info for the milk reorder.

File: .ubidoc.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    rows:
+      - case: "ReceivedOrder"
+
+  - name: "milk-planning-outgoing"
+    termName: "Event"
+    definitionName: "Description"
+    rows:
+      - case: "OrderMilk"
+
+ignored: []
```
Comment (giacomocavalieri) [inline .ubidoc.yml]: You are missing the newline

File: milk-planning/src/main/scala/dev/atedeg/mdm/milkplanning/Actions.scala [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    product: RequestedProduct,
+    stock: Stock,
+    recipeBook: RecipeBook,
+): QuintalsOfMilk =
+  val RequestedProduct(p @ Product(cheeseType, weight), quantity, _) = product
+  val unitsToProduce = (quantity.n: NonNegativeNumber) - stock(p).quantity
+  val gramsToProduce = unitsToProduce * weight.n
+  val quintalsToProduce = gramsToProduce.toDecimal / 100_000
+  val neededQuintals = quintalsToProduce * recipeBook(cheeseType).n
+  neededQuintals.ceil.quintalsOfMilk
```
Comment (giacomocavalieri) [inline milk-planning/src/main/scala/dev/atedeg/mdm/milkplanning/Actions.scala]: I have not checked if it compiles but I think you could go even a little further with pattern matching:
```suggestion
  val RequestedProduct(p @ Product(cheeseType, Grams(weight)), Quantity(requestedUnits), _) = product
  val unitsToProduce = requestedUnits.toNonNegative - stock(p).quantity
  val gramsToProduce = unitsToProduce * weight
  val quintalsToProduce = gramsToProduce.toDecimal / 100_000
  val neededQuintals = quintalsToProduce * recipeBook(cheeseType).n
  neededQuintals.ceil.quintalsOfMilk
```

File: utils/src/main/scala/dev/atedeg/mdm/utils/NumericOps.scala [hunk 4]
```diff
@@ -59,3 +65,15 @@ object Div:
 
   inline given divGen[N](using inst: K0.ProductInstances[Div, N]): Div[N] with
     def div(n1: N, n2: N): N = inst.map2(n1, n2)([n] => (d: Div[n], n1: n, n2: n) => d.div(n1, n2))
+
+given [N, P <: Positive | NonNegative: ValidFor[N]](using C: Ceil[N]): Ceil[N Refined P] with
+  override def toCeil(n: N Refined P): N Refined P = coerce(C.toCeil(n.value))
```
Comment (giacomocavalieri) [inline utils/src/main/scala/dev/atedeg/mdm/utils/NumericOps.scala]: I'd move this in the refined file where we define these instances for the refined types

File: milk-planning/src/test/scala/dev/atedeg/mdm/milkplanning/types/ActionsTest.scala [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import dev.atedeg.mdm.milkplanning.OutgoingEvent.OrderMilk
+import dev.atedeg.mdm.milkplanning.utils.QuintalsOfMilkOps.*
+import dev.atedeg.mdm.milkplanning.utils.QuintalsOfMilkOps.given
+import dev.atedeg.mdm.products.{ CheeseType, Grams, Product }
+import dev.atedeg.mdm.products.Product.*
+import dev.atedeg.mdm.utils.*
+import dev.atedeg.mdm.utils.given
+import dev.atedeg.mdm.utils.monads.*
+
+trait Fixture {
```
Comment (giacomocavalieri) [inline milk-planning/src/test/scala/dev/atedeg/mdm/milkplanning/types/ActionsTest.scala]: @ndido98 called this trait `Mocks` in his PR, we should stick with one name, personally I prefer `Mocks`

File: milk-planning/src/main/scala/dev/atedeg/mdm/milkplanning/Types.scala [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+/**
+ * A quantity of milk expressed in quintals.
+ * @note it must be a [[PositiveDecimal positive decimal number]].
+ * @example `QuintalsOfMilk(1.1)` is a valid weight of 110 kg.
+ * @example `QuintalsOfMilk(-20.5)` is not a valid weight.
+ */
+final case class QuintalsOfMilk(quintals: NonNegativeNumber) derives Plus, Times, Minus
+
+/**
+ * Represent how many [[QuintalsOfMilk quintals of milk]] are needed to produce a given quantity of [[Product product]].
```
Comment (giacomocavalieri) [inline milk-planning/src/main/scala/dev/atedeg/mdm/milkplanning/Types.scala]:
```suggestion
 * A decimal that represents the yield of milk when producing a given [[CheeseType cheese type]]:
 * i.e. to produce `n` quintals of a given [[CheeseType cheese type]], `yield of cheese type * n` [[QuintalsOfMilk quintals of milk]] must be used.
```

File: .ubidoc.yml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+      - type: "Stock"
+      - class: "StockedQuantity"
+      - class: "Quantity"
+      - class: "RequestedProduct"
+
+  - name: "milk-planning-incoming"
+    termName: "Event"
+    definitionName: "Description"
+    rows:
+      - case: "ReceivedOrder"
```
Comment (giacomocavalieri) [inline .ubidoc.yml]: Remember that in the new plugin version (once we merge my open PR) you will have to specify the full case name

**Review by giacomocavalieri — CHANGES_REQUESTED** [review]

File: .ubidoc.yml (hunk 7 above)
Comment (nicolasfara) [reply .ubidoc.yml]: I don't quite understand what you mean. In the new versione I have to specify something like this:
```
rows:
  - case: "IncomingEvent.ReceivedOrder`
```

Comment (giacomocavalieri) [reply .ubidoc.yml]: Yes, exactly; also if you don't like the full name you can specify a custom name:
```yaml 
rows: 
  - case: "IncomingEvent.ReceivedOrder`
    name: "Received Order"
```

**Review by vitlinda — APPROVED** [review]

**Review by giacomocavalieri — APPROVED** [review]

**Pull request merged**